extensions/ql-vscode/src/pure/variant-analysis-filter-sort.ts

```diff
@@ -72,16 +76,22 @@ export function compareWithResults(filterSortState: RepositoriesFilterSortState
   };
 }
 
+function isFilterOnRepositoryIds(filterSortState: RepositoriesFilterSortState | RepositoriesFilterSortStateWithIds | undefined): filterSortState is RepositoriesFilterSortStateWithIds & Required<Pick<RepositoriesFilterSortStateWithIds, 'repositoryIds'>> {
+  return !!filterSortState && 'repositoryIds' in filterSortState && filterSortState.repositoryIds !== undefined && filterSortState.repositoryIds.length > 0;
+}
+
 // These define the behavior for undefined input values
-export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[], filterSortState: RepositoriesFilterSortState | undefined): T[];
-export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | undefined): T[] | undefined;
+export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[], filterSortState: RepositoriesFilterSortState | RepositoriesFilterSortStateWithIds | undefined): T[];
+export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | RepositoriesFilterSortStateWithIds | undefined): T[] | undefined;
 
-export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | undefined): T[] | undefined {
+export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | RepositoriesFilterSortStateWithIds | undefined): T[] | undefined {
```

Comment:
Simplifying the check is useful. 👍 

I seem to remember having this conversation before about not being able to split up this method into two.

Just for the sake of the argument, if we were able to split it, we could have:
- one method to be used by the view (without `repositoryIds`)
- one for everywhere else

E.g. 

```
  export function viewFilterAndSortRepositories<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | undefined): T[] | undefined {
    if (!repositories) {
      return undefined;
    }

    return repositories
      .filter(repo => filterSortState.repositoryIds.includes(repo.repository.id))
      .sort(compareWithResults(filterSortState));
  }

  export function filterAndSortRepositories<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortStateWithIds | undefined): T[] | undefined {
    if (!repositories) {
      return undefined;
    }

    return repositories
      .filter(repo => matchesFilter(repo.repository, filterSortState));
      .sort(compareWithResults(filterSortState));
  }

```

I think the reason this is getting quite intertwined is because the singular method is doing too much so every time we need to touch it, it's going to become harder to change. With smaller methods we'd release ourselves a little bit from that burden.

Reply:
Ok! So I think there's two problems here:

1. We want to make it crystal clear that we shouldn't be using `repositoryIds` in view filtering. 
2. When we do want to filter by repo id, we should only be doing that when there are `repositoryIds` set. Otherwise we should default to the way we do sorting like in the view (by repo name).

I think you can do two things:
1. Leave things as they are as it's doing the job, but maybe shorten the type check from 

```ts
!!filterSortState && 'repositoryIds' in filterSortState && filterSortState.repositoryIds !== undefined && filterSortState.repositoryIds.length > 0
```

to 

```ts
!!filterSortState?.repositoryIds && filterSortState.repositoryIds.length > 0
```

I ** believe ** from testing in the console that those are equivalent.

2. To solve your first problem of making it clear which version should be used in views you could still define a separate `filterAndSortByRepoName` method and replace the call to it in the view:

```ts
export function filterAndSortByRepoName<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortState | undefined): T[] | undefined {
  if (!repositories) {
    return undefined;
  }

  return repositories
    .filter(repo => matchesFilter(repo.repository, filterSortState))
    .sort(compareWithResults(filterSortState));
}
```

Then you could incorporate it into your existing method to make it clear that we fallback to sorting by repo name with repo ids are missing.

```ts
export function filterAndSortRepositoriesWithResults<T extends SortableResult>(repositories: T[] | undefined, filterSortState: RepositoriesFilterSortStateWithIds | undefined): T[] | undefined {
  if (!repositories) {
    return undefined;
  }

  if (isFilterOnRepositoryIds(filterSortState)) {
    return repositories
      .filter(repo => filterSortState.repositoryIds.includes(repo.repository.id))
      .sort(compareWithResults(filterSortState));
  } else {
    filterAndSortByRepoName(repositories, filterSortState);
  }
}
```